extensions/stackdriver/log/logger_test.cc

```diff
@@ -0,0 +1,156 @@
+  auto node_info = nodeInfo();
+  google::logging::v2::WriteLogEntriesRequest req;
+  req.set_log_name(kServerAccessLogName);
+  google::api::MonitoredResource monitored_resource;
+  Common::getMonitoredResource(Common::kContainerMonitoredResource, node_info,
+                               &monitored_resource);
+  req.mutable_resource()->CopyFrom(monitored_resource);
+  auto top_label_map = req.mutable_labels();
+  (*top_label_map)["destination_name"] = node_info.name();
+  (*top_label_map)["destination_workload"] = node_info.workload_name();
+  (*top_label_map)["destination_namespace"] = node_info.namespace_();
+  for (int i = 0; i < log_entry_count; i++) {
+    auto* new_entry = req.mutable_entries()->Add();
+    *new_entry->mutable_timestamp() = TimeUtil::SecondsToTimestamp(0);
+    new_entry->set_severity(::google::logging::type::INFO);
+    auto label_map = new_entry->mutable_labels();
+    (*label_map)["source_name"] = peer_node_info.name();
+    (*label_map)["source_workload"] = peer_node_info.workload_name();
+    (*label_map)["source_namespace"] = peer_node_info.namespace_();
+
+    (*label_map)["request_operation"] = request_info.request_operation;
+    (*label_map)["destination_service_host"] =
+        request_info.destination_service_host;
+    (*label_map)["response_flag"] = request_info.response_flag;
+    (*label_map)["protocol"] = request_info.request_protocol;
+    (*label_map)["destination_principal"] = request_info.destination_principal;
+    (*label_map)["source_principal"] = request_info.source_principal;
+    (*label_map)["service_authentication_policy"] =
+        request_info.mTLS ? "true" : "false";
+  }
+  return req;
+}
+
+}  // namespace
+
+TEST(LoggerTest, TestWriteLogEntry) {
+  auto exporter = std::make_unique<::testing::NiceMock<MockExporter>>();
+  auto exporter_ptr = exporter.get();
+  auto logger = std::make_unique<Logger>(nodeInfo(), std::move(exporter));
+  logger->addLogEntry(requestInfo(), peerNodeInfo());
+  EXPECT_CALL(*exporter_ptr, exportLogs(::testing::_))
+      .Times(1)
+      .WillRepeatedly(::testing::Invoke(
+          [](const google::logging::v2::WriteLogEntriesRequest& request) {
+            auto expected_request = expectedRequest(1);
+            EXPECT_TRUE(MessageDifferencer::Equals(expected_request, request));
+          }));
+  logger->flush();
+}
+
+TEST(LoggerTest, TestWriteLogEntryRotation) {
+  auto exporter = std::make_unique<::testing::NiceMock<MockExporter>>();
+  auto exporter_ptr = exporter.get();
+  auto logger = std::make_unique<Logger>(nodeInfo(), std::move(exporter), 900);
+  EXPECT_CALL(*exporter_ptr, exportLogs(::testing::_))
+      .Times(3)
+      .WillRepeatedly(::testing::Invoke(
+          [](const google::logging::v2::WriteLogEntriesRequest& request) {
+            auto expected_request = expectedRequest(3);
+            EXPECT_TRUE(MessageDifferencer::Equals(expected_request, request));
+          }));
+  for (int i = 0; i < 9; i++) {
```

Comment:
for later: we can improve this test by calling addLogEntry in parallel in multiple threads?

Reply:
I don't think it is necessary. Envoy uses silo thread for filter. We can assume signal thread is always true.